src/ray/common/asio/instrumented_io_context.cc

```diff
@@ -60,6 +60,71 @@ void instrumented_io_context::post(std::function<void()> handler,
   if (!RayConfig::instance().asio_event_loop_stats_collection_enabled()) {
     return boost::asio::io_context::post(std::move(handler));
   }
+  auto stats_handle = RecordStart(name);
+  // References are only invalidated upon deletion of the corresponding item from the
+  // table, which we won't do until this io_context is deleted. Provided that
+  // GuardedHandlerStats synchronizes internal access, we can concurrently write to the
+  // handler stats it->second from multiple threads without acquiring a table-level
+  // readers lock in the callback.
+  // NOTE: We capture the handler stats and global stats shared pointers by value,
+  // ensuring that the reference count will be incremented and that the underlying
+  // structs won't be deallocated until all such handlers finish executing. This allows
+  // the handlers to execute without segfaulting even if the thread that constructed the
+  // io_context object exits without joining the handler threads (which can happen in a
+  // hard exit).
+  boost::asio::io_context::post(
+      [handler = std::move(handler), stats_handle = std::move(stats_handle)]() {
+        RecordExecution(std::move(handler), std::move(stats_handle));
+      });
+}
+
+std::shared_ptr<StatsHandle> instrumented_io_context::RecordStart(const std::string name,
+                                                                  int64_t pad_start_ns) {
+  auto stats = GetOrCreate(name);
+  {
+    absl::MutexLock lock(&(stats->mutex));
+    stats->stats.cum_count++;
+    stats->stats.curr_count++;
+  }
+  return std::make_shared<StatsHandle>(name, absl::GetCurrentTimeNanos() + pad_start_ns,
+                                       stats, global_stats_);
+}
+
+void instrumented_io_context::RecordExecution(std::function<void()> fn,
+                                              std::shared_ptr<StatsHandle> handle) {
+  int64_t start_execution = absl::GetCurrentTimeNanos();
+  // Execute actual handler.
+  fn();
+  int64_t end_execution = absl::GetCurrentTimeNanos();
+  // Update execution time stats.
+  const auto execution_time_ns = end_execution - start_execution;
+  // Update handler-specific stats.
+  {
+    auto stats = handle->handler_stats;
```

Comment:
```suggestion
    auto &stats = handle->handler_stats;
```
Minor, but I think it is safe to just use the reference here.